cvmfs/cache.cc

```diff
@@ -107,178 +111,86 @@ atomic_int32 CallGuard::num_inflight_calls_ = 0;
-             "Alien cache is on NFS.");
+    pos += nbytes;
+    eof = nbytes < 4096;
+    retval = zlib::CompressZStream2Null(buf, nbytes, eof, &strm, &hash_context);
+    if (retval == zlib::kStreamDataError) {
+      zlib::CompressFini(&strm);
+      return -EINVAL;
     }
-  } else {
-    if (!MakeCacheDirectories(cache_path, 0700))
-      return false;
-  }
+  } while (!eof);
 
-  if (FileExists(cache_path + "/cvmfscatalog.cache")) {
-    LogCvmfs(kLogCache, kLogStderr | kLogSyslogErr,
-             "Not mounting on cvmfs 2.0.X cache");
-    return false;
-  }
-
-  int retval = pthread_key_create(&thread_local_storage_, TLSDestructor);
-  assert(retval == 0);
-
-  return true;
-}
-
-
-void Fini() {
-  pthread_mutex_lock(&lock_tls_blocks_);
-  for (unsigned i = 0; i < tls_blocks_->size(); ++i)
-    CleanupTLS((*tls_blocks_)[i]);
-  pthread_mutex_unlock(&lock_tls_blocks_);
-  pthread_key_delete(thread_local_storage_);
-  delete cache_path_;
-  delete queues_download_;
-  delete tls_blocks_;
-  cache_path_ = NULL;
-  queues_download_ = NULL;
-  tls_blocks_ = NULL;
-}
-
-
-CacheModes GetCacheMode() {
-  return cache_mode_;
-}
-
-void TearDown2ReadOnly() {
-  cache_mode_ = kCacheReadOnly;
-  CallGuard::Drainout();
-  quota::Fini();
-  unlink(("running." + *cvmfs::repository_name_).c_str());
-  LogCvmfs(kLogCache, kLogSyslog, "switch to read-only cache mode");
-  SetLogMicroSyslog("");
+  zlib::CompressFini(&strm);
+  if (retval != zlib::kStreamEnd)
+    return -EINVAL;
+  shash::Final(hash_context, id);
+  return 0;
 }
 
 
 /**
- * Transforms a catalog entry into a name for local cache.
- *
- * @param[in] id content hash of the catalog entry.
- * \return Absolute path in local cache.
+ * Commits the memory blob buffer to the given chunk id.  No checking!
+ * The hash and the memory blob need to match.
  */
-static inline string GetPathInCache(const shash::Any &id) {
-  return *cache_path_ + "/" + id.MakePathWithoutSuffix();
-}
-
-
-/**
- * Transform a catalog entry into a temporary name in txn-directory.
- *
- * @param[in] id content hash of the catalog entry.
- * \return Absolute path in local cache txn-directory.
- */
-static inline string GetTempName()
+bool CacheManager::CommitFromMem(
+  const shash::Any &id,
+  const unsigned char *buffer,
+  const uint64_t size,
+  const string &description)
 {
-  return *cache_path_ + "/txn/" + "fetchXXXXXX";
-}
-
-
-/**
- * Tries to open a catalog entry in local cache.
- *
- * @param[in] id content hash of the catalog entry.
- * \return A file descriptor if file is in cache.  Error code of open() else.
- */
-int Open(const shash::Any &id) {
-  const string path = GetPathInCache(id);
-  int result = ::open(path.c_str(), O_RDONLY);
-
-  if (result >= 0) {
-    LogCvmfs(kLogCache, kLogDebug, "hit %s", path.c_str());
-    // platform_disable_kcache(result);
-  } else {
-    result = -errno;
-    LogCvmfs(kLogCache, kLogDebug, "miss %s (%d)", path.c_str(), result);
+  void *txn = alloca(this->SizeOfTxn());
```

Comment:
Is there a reason that a `Transaction` is not represented as a struct or maybe a hierarchy of structs? Wouldn't that make all this transaction allocation transparent (`SizeOfTxn()`, `alloca()` and `void*`)?